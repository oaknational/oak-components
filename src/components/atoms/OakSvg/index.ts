export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
