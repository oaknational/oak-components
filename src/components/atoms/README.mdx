# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
