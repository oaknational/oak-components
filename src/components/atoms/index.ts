export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
