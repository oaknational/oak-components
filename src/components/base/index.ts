export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakBaseCard";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
