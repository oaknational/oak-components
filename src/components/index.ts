export * from "./base";
// export * from "./ui"; // uncomment this line once there are components in this folder
// export * from "./integrated"; // uncomment this line once there are components in this folder
