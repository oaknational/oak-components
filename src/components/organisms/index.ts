export * from "./pupil";
export * from "./teacher";
export * from "./OakSubjectIcon";
export * from "./OakPrimaryNavItem";
export * from "./OakPrimaryNav";
export * from "./OakCookieBanner";
export * from "./OakCookieSettingsModal";
export * from "./OakCookieConsentProvider";
export * from "./OakCookieConsent";
