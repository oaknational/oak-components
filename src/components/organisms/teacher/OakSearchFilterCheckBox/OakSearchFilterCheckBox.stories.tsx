import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakSearchFilterCheckBox } from "./OakSearchFilterCheckBox";

const meta: Meta<typeof OakSearchFilterCheckBox> = {
  component: OakSearchFilterCheckBox,
  tags: ["autodocs"],
  title: "components/organisms/teacher/OakSearchFilterCheckBox",
  parameters: {
    backgrounds: {
      default: "light",
    },
    controls: {
      include: ["disabled", "defaultChecked", "displayValue", "value", "icon"],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakSearchFilterCheckBox>;

export const Default: Story = {
  render: (args) => <OakSearchFilterCheckBox {...args} />,
  args: {
    id: "checkbox-test-default",
    value: "a test value",
    displayValue: "Art and design",
    icon: "subject-art",
    "aria-label": "Art and design",
  },
  parameters: {
    controls: {
      include: ["disabled", "defaultChecked", "displayValue", "value", "icon"],
    },
  },
};
