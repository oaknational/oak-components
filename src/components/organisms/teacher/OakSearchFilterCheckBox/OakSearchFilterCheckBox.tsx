import React, { useRef } from "react";
import styled from "styled-components";

import { OakFlex, OakIcon, OakIconName } from "@/components/atoms";
import { InternalCheckBoxLabelHoverDecor } from "@/components/atoms/InternalCheckBoxLabel";
import {
  BaseCheckBoxProps,
  InternalCheckBox,
} from "@/components/atoms/InternalCheckBox/InternalCheckBox";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseBorder } from "@/styles/helpers/parseBorder";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";
import { parseColorFilter } from "@/styles/helpers/parseColorFilter";

// Converted to styled-component so it can be used in '&:checked:not(:disabled) + ${StyledOakIcon}' to change svg color.
const StyledOakIcon = styled(OakIcon)``;

const StyledInternalCheckBox = styled(InternalCheckBox)<{
  $keepIconColor?: boolean;
}>`
  position: absolute;
  opacity: 0;
  pointer-events: none;

  &:checked:not(:disabled) + ${StyledOakIcon} {
    filter: ${(props) =>
      props.$keepIconColor ? "none" : parseColorFilter("white")};
  }

  &:checked:not(:disabled) {
    border: ${parseBorder("border-solid-l")};
    border-color: ${parseColor("border-primary")};
  }

  &:checked:disabled {
    border: ${parseBorder("border-solid-l")};
    border-color: ${parseColor("text-disabled")};
  }

  &:hover:not(:disabled) {
    background: ${parseColor("bg-primary")};
  }
`;

const StyledFlexBox = styled(OakFlex)`
  &:has(input:not(:disabled)) {
    cursor: pointer;
  }

  &:has(input:disabled) {
    pointer-events: none;
    cursor: none;
  }

  &:hover:has(input:not(:disabled)) ${InternalCheckBoxLabelHoverDecor} {
    text-decoration: underline;
  }

  &:hover:has(input:not(:disabled)) {
    background-color: ${parseColor("bg-neutral")};
  }

  &:focus-within {
    box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")},
      ${parseDropShadow("drop-shadow-centered-grey")};
  }

  &:has(input:checked:not(:disabled)) {
    background-color: ${parseColor("black")};
    color: ${parseColor("white")};
  }
`;

export type OakSearchFilterCheckBoxProps = Omit<
  BaseCheckBoxProps,
  "defaultChecked"
> & {
  innerRef?: React.RefObject<HTMLInputElement>;
  displayValue: string;
  icon?: OakIconName;
  keepIconColor?: boolean;
};

/**
 * A checkbox for search filters.
 *
 * Takes a displayValue and optional icon, for subject icons 'subject-[subjectSlug]' should be used.
 *
 * ## Events
 * The following callbacks are available for tracking focus events:
 *
 * ### onChange
 * onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
 *
 * ### onFocus
 *   onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
 *
 * ### onBlur
 *    onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
 *
 * ### onHovered
 *  `onHovered?: (id, value, duration: number) => void;`<br>
 *  called after a mouseEnter and mouseLeave event has happened
 *
 */
export const OakSearchFilterCheckBox = (
  props: OakSearchFilterCheckBoxProps,
) => {
  const {
    id,
    value,
    disabled,
    innerRef,
    displayValue,
    icon,
    keepIconColor,
    ...rest
  } = props;

  const defaultRef = useRef<HTMLInputElement>(null);
  const inputRef = innerRef ?? defaultRef;

  const handleContainerClick = (
    e:
      | React.MouseEvent<HTMLDivElement>
      | React.MouseEvent<HTMLInputElement>
      | React.MouseEvent<HTMLLabelElement>,
  ) => {
    const inputId = (e.target as HTMLInputElement).id;

    if (inputId !== id) {
      inputRef.current?.click();
    }
  };

  return (
    <OakFlex $minHeight={"all-spacing-8"} $position={"relative"}>
      <StyledFlexBox
        $borderRadius={"border-radius-s"}
        $borderColor={"border-neutral-lighter"}
        $ba="border-solid-s"
        $background={"white"}
        onClick={handleContainerClick}
        $ph={"inner-padding-s"}
        $pv={"inner-padding-ssx"}
        $gap={"space-between-sssx"}
      >
        <StyledInternalCheckBox
          id={id}
          value={value}
          disabled={disabled}
          ref={inputRef}
          $keepIconColor={keepIconColor}
          {...rest}
        />
        {icon && <StyledOakIcon alt="" iconName={icon} />}
        <InternalCheckBoxLabelHoverDecor
          pointerEvents="none"
          htmlFor={id}
          $font={"heading-7"}
          disabled={disabled}
        >
          {displayValue}
        </InternalCheckBoxLabelHoverDecor>
      </StyledFlexBox>
    </OakFlex>
  );
};
