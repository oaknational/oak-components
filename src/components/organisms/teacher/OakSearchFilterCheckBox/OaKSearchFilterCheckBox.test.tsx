import React, { createRef } from "react";
import "@testing-library/jest-dom";
import { create } from "react-test-renderer";
import { fireEvent } from "@testing-library/react";

import { OakSearchFilterCheckBox } from "./OakSearchFilterCheckBox";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { OakThemeProvider } from "@/components/atoms";
import { oakDefaultTheme } from "@/styles";

describe("OakSearchFilterCheckBox", () => {
  it("renders a checkbox", () => {
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        data-testid="test-1"
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    expect(getByRole("checkbox")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <OakSearchFilterCheckBox
          id="checkbox-1"
          value="Option 1"
          icon={"subject-history"}
          displayValue="Option 1"
        />
        ,
      </OakThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has a label", () => {
    const { getByLabelText } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        displayValue="Option 1"
        value="1"
        icon={"subject-history"}
      />,
    );
    expect(getByLabelText("Option 1")).toBeInTheDocument();
  });

  it("calls onChange when clicked", () => {
    const onChange = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        onChange={onChange}
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    getByRole("checkbox").click();
    expect(onChange).toHaveBeenCalled();
  });

  it("calls onFocus when focused", () => {
    const onFocus = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        onFocus={onFocus}
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    getByRole("checkbox").focus();
    expect(onFocus).toHaveBeenCalled();
  });

  it("calls onBlur when blurred", () => {
    const onBlur = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        onBlur={onBlur}
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    getByRole("checkbox").focus();
    getByRole("checkbox").blur();
    expect(onBlur).toHaveBeenCalled();
  });

  it("calls onHovered when mouse entered and left", () => {
    const onHovered = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        onHovered={onHovered}
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    fireEvent.mouseEnter(getByRole("checkbox"));
    fireEvent.mouseLeave(getByRole("checkbox"));
    expect(onHovered).toHaveBeenCalled();
  });

  it("is not clickable when disabled", () => {
    const onChange = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        onChange={onChange}
        disabled
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    getByRole("checkbox").click();
    expect(onChange).not.toHaveBeenCalled();
  });

  it("can be checked and unchecked using a ref", () => {
    const ref = createRef<HTMLInputElement>();
    const { getByRole } = renderWithTheme(
      <OakSearchFilterCheckBox
        id="checkbox-1"
        value="Option 1"
        innerRef={ref}
        icon={"subject-history"}
        displayValue="Option 1"
      />,
    );
    ref.current?.click();
    expect(getByRole("checkbox")).toBeChecked();
    ref.current?.click();
    expect(getByRole("checkbox")).not.toBeChecked();
  });
});
