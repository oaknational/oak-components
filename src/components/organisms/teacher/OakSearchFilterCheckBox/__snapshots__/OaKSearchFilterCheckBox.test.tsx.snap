// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakSearchFilterCheckBox matches snapshot 1`] = `
[
  .c0 {
  position: relative;
  width: 100%;
  font-family: Lexend,sans-serif;
}

.c2 {
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  background: #ffffff;
  border-color: #cacaca;
  border-radius: 0.25rem;
  border-style: solid;
  font-family: Lexend,sans-serif;
}

.c2:hover {
  cursor: pointer;
}

.c7 {
  position: relative;
  width: 2rem;
  min-width: 2rem;
  height: 2rem;
  min-height: 2rem;
  font-family: Lexend,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
}

.c8 {
  object-fit: contain;
}

.c9 {
  font-family: Lexend,sans-serif;
  font-weight: 600;
  font-size: 1rem;
  line-height: 1.25rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  pointer-events: none;
  cursor: pointer;
}

.c5 {
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  margin: 0;
  outline: none;
  border: 0.125rem solid;
  border-color: #222222;
  border-radius: 0.125rem;
}

.c5:checked {
  background: #222222;
}

.c5:disabled {
  pointer-events: none;
}

.c6 {
  position: absolute;
  opacity: 0;
  pointer-events: none;
}

.c6:checked:not(:disabled) {
  border: 0.188rem solid;
  border-color: #222222;
}

.c6:checked:disabled {
  border: 0.188rem solid;
  border-color: #808080;
}

.c4:has(input:not(:disabled)) {
  cursor: pointer;
}

.c4:has(input:disabled) {
  pointer-events: none;
  cursor: none;
}

@media (hover:hover) {
  .c11:hover {
    -webkit-text-decoration: underline;
    text-decoration: underline;
  }
}

@media (hover:hover) {
  .c5:hover:not(:disabled) {
    border-color: #222222;
  }
}

@media (hover:hover) {
  .c6:hover:not(:disabled) {
    background: #ffffff;
  }
}

@media (hover:hover) {
  .c4:hover:has(input:not(:disabled)) {
    background-color: #f2f2f2;
  }

  .c4:focus-within {
    box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%), 0 0 0 0.3rem rgba(87,87,87,100%);
  }

  .c4:has(input:checked:not(:disabled)) {
    background-color: #222222;
    color: #ffffff;
  }
}

<div
    className="c0 c1"
  >
    <div
      className="c2 c3 c4"
      onClick={[Function]}
    >
      <input
        className="c5 c6"
        id="checkbox-1"
        name="checkbox-1"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        type="checkbox"
        value="Option 1"
      />
      <div
        className="c7 "
      >
        <img
          alt="subject-history"
          className="c8"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/subject-icons/history.svg"
          style={
            {
              "bottom": 0,
              "color": "transparent",
              "height": "100%",
              "left": 0,
              "objectFit": undefined,
              "objectPosition": undefined,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": "100%",
            }
          }
        />
      </div>
      <label
        className="c9 c10 c11"
        htmlFor="checkbox-1"
        pointerEvents="none"
      >
        Option 1
      </label>
    </div>
  </div>,
  ",",
]
`;
